be/src/tools/meta_tool.cpp

```diff
@@ -433,6 +439,80 @@ void show_segment_footer(const std::string& file_name) {
     std::cout << json_footer << std::endl;
 }
 
+// This function will check the consistency of tablet meta and segment_footer
+// #issue 5415
+void check_meta_consistency(DataDir* data_dir) {
+    std::unique_ptr<MemTracker> mem_tracker = std::make_unique<MemTracker>();
+    starrocks::GlobalTabletSchemaMap::Instance()->set_mem_tracker(mem_tracker.get());
+    std::vector<int64_t> tablet_ids;
+    int64_t table_id = FLAGS_table_id;
+    auto check_meta_func = [data_dir, &tablet_ids, table_id](int64_t tablet_id, int32_t schema_hash,
+                                                             std::string_view value) -> bool {
+        starrocks::TabletMetaSharedPtr tablet_meta(new TabletMeta());
+        // if deserialize failed, skip it
+        if (Status st = tablet_meta->deserialize(value); !st.ok()) {
+            return true;
+        }
+        // tablet is not belong to the table, skip it
+        if (tablet_meta->table_id() != table_id) {
+            return true;
+        }
+        std::string tablet_path = data_dir->path() + starrocks::DATA_PREFIX;
+        tablet_path = starrocks::path_util::join_path_segments(tablet_path, std::to_string(tablet_meta->shard_id()));
+        tablet_path = starrocks::path_util::join_path_segments(tablet_path, std::to_string(tablet_meta->tablet_id()));
+        tablet_path = starrocks::path_util::join_path_segments(tablet_path, std::to_string(tablet_meta->schema_hash()));
+
+        auto& tablet_schema = tablet_meta->tablet_schema();
+        const std::vector<starrocks::TabletColumn>& columns = tablet_schema.columns();
+
+        for (const auto& rs : tablet_meta->all_rs_metas()) {
+            for (int64_t seg_id = 0; seg_id < rs->num_segments(); ++seg_id) {
+                std::string seg_path =
+                        strings::Substitute("$0/$1_$2.dat", tablet_path, rs->rowset_id().to_string(), seg_id);
+                auto res = starrocks::FileSystem::Default()->new_random_access_file(seg_path);
+                if (!res.ok()) {
+                    continue;
+                }
+                auto seg_file = std::move(res).value();
+                starrocks::SegmentFooterPB footer;
+                auto status = get_segment_footer(seg_file.get(), &footer);
+                if (!status.ok()) {
+                    continue;
+                }
+
+                std::unordered_map<uint32_t, int32_t> column_unique_id_to_type;
+                for (uint32_t ordinal = 0; ordinal < footer.columns().size(); ++ordinal) {
+                    const auto& column_pb = footer.columns(ordinal);
+                    column_unique_id_to_type.emplace(column_pb.unique_id(), column_pb.type());
+                }
+                for (uint32_t col_id = 0; col_id < columns.size(); ++col_id) {
+                    uint32_t unique_id = columns[col_id].unique_id();
+                    starrocks::FieldType type = columns[col_id].type();
```

Comment:
> You should check the type equality like varchar/decimal length.

Done. But we can only check column length if column type is varchar. Because we don't save precision in the footer if the column type is decimal